lib/storages/payments_database.ts

```diff
@@ -1,48 +1,77 @@
-import Engine from '../engines/engine'
 import { ChannelId } from '../channel'
-import Payment from '../Payment'
-import util = require('ethereumjs-util')
+import Payment, { PaymentJSON } from '../Payment'
+import EngineMongo from '../engines/engine_mongo'
+import pify from '../util/pify'
+import Engine from '../engines/engine'
+import EngineNedb from '../engines/engine_nedb'
+import EnginePostgres from '../engines/engine_postgres'
+import { namespaced } from '../util/namespaced'
 
-const namespaced = (namespace: string | null | undefined, kind: string): string => {
-  let result = kind
-  if (namespace) {
-    result = namespace + ':' + kind
-  }
-  return result
+// linter false positive below
+/* tslint:disable */
+import BigNumber from '../bignumber'
+/* tslint:enable */
+
+export default interface PaymentsDatabase {
+  save (token: string, payment: Payment): Promise<void>
+
+  firstMaximum (channelId: ChannelId | string): Promise<Payment | null>
+
+  findByToken (token: string): Promise<Payment | null>
 }
 
-/**
- * Database layer for payments.
- */
-export default class PaymentsDatabase {
+export abstract class AbstractPaymentsDatabase<T extends Engine> implements PaymentsDatabase {
   kind: string
-  engine: Engine
 
-  constructor (engine: Engine, namespace: string | null) {
+  engine: T
+
+  constructor (engine: T, namespace: string | null) {
     this.kind = namespaced(namespace, 'payment')
     this.engine = engine
   }
 
+  inflatePayment (json: PaymentJSON): Payment|null {
+    if (!json) {
+      return null
+    }
+
+    return new Payment({
+      channelId: json.channelId,
+      sender: json.sender,
+      receiver: json.receiver,
+      price: new BigNumber(json.price),
+      value: new BigNumber(json.value),
+      channelValue: new BigNumber(json.channelValue),
+      v: Number(json.v),
+      r: json.r,
+      s: json.s,
+      meta: json.meta,
+      contractAddress: json.contractAddress,
+      token: json.token
+    })
+  }
+
+  abstract save (token: string, payment: Payment): Promise<void>
+
+  abstract firstMaximum (channelId: ChannelId | string): Promise<Payment | any>
+
+  abstract findByToken (token: string): Promise<Payment | any>
+}
+
+/**
+ * Database layer for payments.
+ */
+export class MongoPaymentsDatabase extends AbstractPaymentsDatabase<EngineMongo> {
   /**
    * Save payment to the database, to check later.
    */
   save (token: string, payment: Payment): Promise<void> {
-    let document = {
-      kind: this.kind,
-      token: token,
-      channelId: payment.channelId,
-      value: util.bufferToHex(util.toBuffer(payment.value.toString())),
```

Comment:
Rather than storing values as hex, I figured it would be clearer to store them as strings of numbers as per BigNumber's `toString()` method. This confers a number of benefits:

1. Reduced overhead (i.e., no more need to convert the string value to a buffer, then that buffer to hex)
2. The database can order numeric strings automatically.
3. We're already storing a version of the BigNumber's string representation anyway.